Add SSL support to mug

arnu515 opened the pull request:
Hi! I've added some SSL support to mug using Erlang's [`ssl`](https://www.erlang.org/doc/apps/ssl/) module. It lives in the `mug/ssl` module and has an API very similar to the existing TCP API.

Here's an example of connecting to an SSL host:

```gleam
import gleam/bit_array
import gleam/io
import mug
import mug/ssl

pub fn main() {
  let assert Ok(socket) =
    mug.new("localhost", port: 12345)  // connect to a TCP/SSL echo server
    |> ssl.with_ssl()
    |> ssl.with_cacerts(ssl.NoVerification)  // don't verify if certs are valid (useful for local debugging)
    |> ssl.connect()

  let assert Ok(Nil) = ssl.send(socket, <<"Hello, Joe!\n":utf8>>)
  let assert Ok(packet) = ssl.receive(socket, timeout_milliseconds: 100)
  io.println(bit_array.to_string(packet))  // "Hello, Joe!\n"
}
```

There are still a few bits and bobs missing, like the ability to upgrade from a regular TCP connection to an encrypted TLS one. There is also an unimplemented function for adding the [certs_keys](https://www.erlang.org/doc/apps/ssl/ssl.html#t:common_option_cert/0) option since I'm not able to figure out how to do it right now (I'll give it a try soon, I'm new to both Gleam and Erlang).

I've added a few tests in `test/mug_ssl_test.gleam` which are built off of `test/mug_test.gleam`. One of them isn't currently working because I'm not able to figure out some ca certificate nonsense (hence, it has been commented out). 

Please let me know if any changes have to be done, and I'd request you check my doc comments for any grammatical errors, since English is not my first language. Thanks!

arnu515 commented:
Hi! I've added the ability to upgrade a TCP connection to a TLS connection via `upgrade` and `upgrade3`.

There are still two things left:
- Supporting `certs_keys`
- Fixing that CA certificate test

(closed by mistake, sorry!)

lpil commented on src/mug/ssl.gleam:
```diff
@@ -0,0 +1,334 @@
+
+type DoNotLeak
+
+pub type SSLConnectionOptions {
```
This would be "s s l connection options"
```suggestion
pub type SslConnectionOptions {
```

lpil commented on src/mug/ssl.gleam:
```diff
@@ -0,0 +1,334 @@
+import gleam/bytes_builder.{type BytesBuilder}
```
SSL and not TLS?

arnu515 replied:
It's named after the erlang module `ssl`. Their explanation for the name `ssl` is:

> The application's name is still SSL because the first versions of the TLS protocol were named SSL (Secure Socket Layer). However, no version of the old SSL protocol is supported by this application.
> 
> — *From the [erlang docs](https://www.erlang.org/doc/apps/ssl/ssl.html)*

I can change it to `tls` if you wish!

lpil replied:
We should use the names that makes more sense for the actual domain rather than naming it after the tech debt of an implementation detail 👍

arnu515 replied:
Alright, I'll change the module's name to `tls`!

lpil commented on src/mug/ssl.gleam:
```diff
@@ -0,0 +1,334 @@
+  SSLConnectionOptions(..options, cacerts: cacerts)
+}
+
+// TODO: Support certs_keys option of [common_option_cert](https://www.erlang.org/doc/apps/ssl/ssl.html#t:common_option_cert/0).
+fn with_certs_keys() {
+  todo as "not implemented"
+}
```
This code is unfinished

lpil commented on src/mug/ssl.gleam:
```diff
@@ -0,0 +1,334 @@
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+pub fn upgrade3(
```
Rename this to `upgrade` and remove the other please.

lpil commented on src/mug/ssl.gleam:
```diff
@@ -0,0 +1,334 @@
+  /// der-encoded CA certificates as BitArrays. A pem-file cannot be used with this option
+  /// because der-encoded certs have a higher priority over pem-encoded cert files, and the
+  /// `cacerts_get/0` function returns der-encoded certs.
+  WithSystemCertificatesAnd(cacerts: List(BitArray))
```
This name seems not in keeping with the others, and uses an usual format. What else could we call this?

arnu515 replied:
Maybe I can drop the `With`? I'm all out of ideas :shrug:

lpil replied:
Looking at this again I think the whole type may need another look. It says it's cacerts, but it has a variant called `NoVerification`, which isn't a cacert.

Perhaps something like this:

```gleam
pub type Security {
  VerifyTls(cacerts: List(CaCerts))
  DangerouslySkipTlsVerification
}

pub type CaCerts {
  SystemCaCertificates
  CustomDerCaCertificates(cacerts: List(BitArray))
  CustomPemCaCertificates(cacertfile: String)
}
```

Then folks can provide whatever cacerts they wish

arnu515 replied:
This is much better!

arnu515 replied:
~~Using an Option would probably be better than using `Security`, which imho is a poorly named type (since security can mean anything, not just something to do with certificates).~~ I can add some convenience methods like `no_verification`, which sets the certificates to `None`, and `use_system_certificates`, which sets the certificates to the system CA certs.

Edit: I've decided against using an option. This is how I'll be doing it (similar to yours above), but with added support for certs_keys:
```gleam
pub type SslConnectionOptions {
  SslConnectionOptions(
    host: String,
    port: Int,
    timeout: Int,
    certificates: Certificates,
  )
}

/// The certificates to use 
pub type Certificates {
  /// Uses these CA certificates and regular certificates to verify the server's certificate.
  ///
  /// The `use_system_cacerts` option makes mug use the system's CA certificates, which are 
  /// usually set to a group of competent CAs who sign most of the web's certificates.  
  /// You may specifiy your own CA certificates with the `cacerts` option, and custom
  /// certificates and their keys with the `certs_keys` option.
  Certificates(
    use_system_cacerts: Bool,
    cacerts: Option(CaCertificates),
    certs_keys: List(CertsKeys),
  )
  /// Do not verify certificates. While this does allow you to use self-signed certificates,
  /// it is highly recommended to not skip verification, and add a custom CA instead.
  NoVerification
}

/// The CA certificates to use
pub type CaCertificates {
  /// Use these der-encoded certificates as CA certificates.
  DerEncodedCaCertificates(cacerts: List(BitArray))
  /// Path to a pem-encoded file which contains CA certificates.
  PemEncodedCaCertificates(cacertfile: String)
}

pub type CertsKeys {
  /// A list of DER-encoded certificates and their corresponding key.
  DerEncodedCertsKeys(cert: List(BitArray), key: Key)
  /// Path to a file containing PEM-encoded certificates and their key, with an optional
  /// password associated with the file containing the key.
  PemEncodedCertsKeys(
    certfile: String,
    keyfile: String,
    password: Option(BitArray),
  )
}

pub type Key {
  /// A der-encoded key.  
  /// `alg` is one of 'RSAPrivateKey' | 'DSAPrivateKey' | 'ECPrivateKey' | 'PrivateKeyInfo'.
  DerEncodedKey(alg: String, key: BitArray)
}
```

lpil replied:
It must not be an option, it must use the descriptive names. I am open to a different name for the type if you have a better one.

lpil replied:
Please use the name I used above that makes it extremely clear when you are disabling tls verification. It is vital that someone unfamiliar with this area understands what they are doing.

The field and type should not be called "certificates" as it has a variant which is not certificates.

arnu515 replied:
I've changed the type name and field to verification, but I feel that NoVerification should stick, since that's what many other languages/libraries use. I've left some thoughts in the later inline comment you've put.

I'll mark this resolved since the code is outdated. Let's continue discussion in the inline comment occuring later on.

lpil commented on src/mug/ssl.gleam:
```diff
@@ -0,0 +1,334 @@
+///
+pub fn connect(options: SSLConnectionOptions) -> Result(Socket, Error) {
+  let host = charlist.from_string(options.host)
+  use _ <- result.try(ssl_start())
```
It's not the job of library code to start OTP applications. Please remove this.

Instead the library could depend on the `ssl` application, or the documentation could tell the user to depend on the `ssl` application.

arnu515 replied:
Sorry, I don't understand what you mean. Is the user supposed to start the SSL application themselves? In that case I guess it'd be best to expose a `start()` and `stop()` function. `start()` could also return the default `SslConnectionOptions`, which would mean I'd be able to remove `with_ssl`. I guess I could do it in a function that accepts a callback so that users could use `use` too.

Please let me know your thoughts on this @lpil.

lpil replied:
Application dependencies are declared in `.app/mix.exs/gleam.toml` and then the init system brings up applications in the required order. Outside of the Erlang shell you should never be starting and stopping applications manually.

arnu515 replied:
Ah I see, that's really cool.

I was looking through glisten's source code, and it looks like the library starts the application there:

```gleam
// src/glisten.gleam

/// Start the SSL server with the given handler on the provided port.  The key
/// and cert files must be provided, valid, and readable by the current user.
pub fn serve_ssl(
  handler: Handler(user_message, data),
  port port: Int,
  certfile certfile: String,
  keyfile keyfile: String,
) -> Result(Subject(supervisor.Message), StartError) {
  start_ssl_server(handler, port, certfile, keyfile)
  |> result.map(get_supervisor)
}

pub fn start_ssl_server(
  handler: Handler(user_message, data),
  port port: Int,
  certfile certfile: String,
  keyfile keyfile: String,
) -> Result(Server, StartError) {
  let assert Ok(_nil) = ssl.start()
  // ...
}
```

I was just wondering why glisten doesn't ask the user to depend on the ssl application.

lpil replied:
Very odd, I don't know why that is.

@rawhat Why is it that Glisten is working outside of the OTP init system here? https://github.com/rawhat/glisten/blob/32c3b97f20d3699ee6b6759b2747246a83286736/src/glisten.gleam#L351

rawhat replied:
I was concerned about it failing to start on certain systems, and this was my not very good solution that I didn't think too much about.

I think it's probably totally reasonable to just add it to `extra_applications` for `glisten`.  I've pushed a change for that.  The next release should be properly conformant 🙏🏻

arnu515 replied:
What about the usecases without ssl? That'll result in the ssl application being started unnecessarily.

lpil replied:
What problem would you foresee arising from that?

arnu515 replied:
I'm new to this, so I may be wrong, but if the ssl application is turned on even when the user is not using the TLS features of glisten, and the ssl app crashes for whatever reason, then it'll bring down the entire runtime, right?

lpil replied:
It would not, no. Erlang's process isolation and supervision system is used to prevent this. The `ssl` application is part of Erlang, it's probably the most trustworthy of all the application.

arnu515 replied:
Alright, I'll add `ssl` to the `extra_applications` of mug.

lpil commented on test/mug_ssl_test.gleam:
```diff
@@ -0,0 +1,169 @@
+//     |> ssl.connect()
+//     |> io.debug()
+//   Nil
+// }
```
Still to do, as you said :)

arnu515 replied:
This is done!

lpil commented on test/mug_test.gleam:
```diff
@@ -8,20 +8,29 @@ import gleeunit
+    )
 
   gleeunit.main()
+  // process.sleep_forever()
```
Remove this please 🙏

arnu515 replied:
Ah that was added by me during testing and I forgot to remove it :sweat_smile:

lpil reviewed (commented):
This is looking really nice! Very clean!! I've left some minor notes inline.

When you are ready for a review please un-draft this PR. Thank you!

arnu515 commented:
There are still two failing test cases - both related to custom certificates. Everything else should (hopefully) be in shape!

arnu515 commented:
Okay. I've narrowed the issue down:

The tests create a self-signed CA certificate and another certificate signed with the CA for TLS.

As far as I understand, erlang's `ssl` module does not allow me to supply CAs which are self-signed. Go's `tls` library does allow me to pass a self-signed CA certificate, however, and I don't quite understand the reason why erlang does not. There was a discussion in [this issue](https://github.com/erlang/otp/issues/8057).

I suggest I remove the failing test case.

arnu515 commented:
I've merged the latest changes from main, but looks like GitHub isn't showing it here for some reason

lpil commented on README.md:
```diff
@@ -33,4 +33,81 @@ pub fn main() {
+
+pub fn main() {
+  // Either depend on the `ssl` application in `gleam.toml`, or run:
+  mug.start()
```
This isn't right, you should not be starting anything here.

arnu515 replied:
I see.

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+
+  // For connect only
+  /// An invalid option was passed
+  Options(opt: Dynamic)
```
What's this dynamic?

arnu515 replied:
The [docs](https://www.erlang.org/doc/apps/ssl/ssl.html#connect/4) types it as any, and doesn't explain what it's supposed to be, but from conventions I assume that this is raised if a passed option is invalid.

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+  NoApplicationProtocol
+}
+
+pub type Error {
```
This is large a duplication of the existing error type, let's merge them into one.

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+  )
+  /// Do not verify certificates. While this does allow you to use self-signed certificates,
+  /// it is highly recommended to not skip verification, and add a custom CA instead.
+  NoVerification
```
This is not a certificate and the name isn't scary enough. Rename this type and the variants please.

arnu515 replied:
I have renamed the type to `VerificationMethod`, but I do not feel that it is necessary to rename `NoVerification` into something dangerous. From the issue discussion I've linked above, and from the things mentioned in the `ssl` documentation, there seems to be an "encouragement" of some sorts to disable verification, instead of using custom self-signed certificates.

Please correct me if I've misunderstood.

lpil replied:
This will not be merged without this change.

The API must very strongly discourage the programmer from disabling TLS verification. There has recently been multiple instances of vulnerabilities due to inexperienced developers disabling TLS verification without understanding, and it is our obligation as library creators to prevent that.

The style of Erlang's documentation and APIs are of no consequence, we do not take them into account when designing our APIs.

arnu515 replied:
I see. Fixed.

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+      }),
+    ),
+  ]
+  |> fn(opts) {
```
Never pipe into a function please

arnu515 replied:
I do not know why I did it like that haha. Fixed.

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+  #(List(BitArray), #(Dynamic, Dynamic, Dynamic))
+
+@external(erlang, "public_key", "cacerts_get")
+fn get_system_cacerts() -> CombinedCert
```
This function will raise if it cannot load the certs. If it is to be used then that needs to be converted into a result try and returned to the user.

arnu515 replied:
I've caught the error using `gleam_erlang.rescue` and propagated it to the user.

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+      Cacertfile,
+      dynamic.from(string.to_utf_codepoints(cacerts)),
+    )
+    True, None -> #(Cacerts, dynamic.from(get_system_cacerts()))
```
Why do we load the certs rather than have the SSL application do that? It would be able to cache them, for example.

arnu515 replied:
I don't quite understand what you mean. Does the SSL application automatically load system cacerts?

arnu515 replied:
And caching is done by erlang's ssl module automatically

lpil commented on src/mug/tls.gleam:
```diff
@@ -0,0 +1,630 @@
+import gleam/bytes_builder.{type BytesBuilder}
```
This module seems to be almost exactly the same as the existing module. We could merge them into one rather than having duplicate functions for the programmer to pick between.

lpil reviewed (commented):
Great work, thank you! I've left notes on the API largely. The TLS side I'm less qualified to speak about, but it looks good from what I can tell.

arnu515 commented:
For merging them into one, I wonder how common methods like `send` and `recv` can be realised. Do I add prefixes, like `ssl_send`, or do I have some kind of record that holds metadata and match on it?

I'd request you to give me another review of the code, and bring to my notice any other problems, which I'll fix before doing this refactoring.

lpil commented:
That sounds good! I presume we cannot cheaply and reliably introspect the connection to determine what mode it is in.

arnu515 commented:
Hi! I've merged the two files into one. There's a unified API now for both TCP and TLS.

A simple example:

```gleam
let assert Ok(socket) =
  mug.new("echo.example.com", port: 443)
  |> mug.timeout(milliseconds: 10_000)
  |> mug.with_tls()  // enable TLS!
  |> mug.connect()
let assert Ok(Nil) = mug.send(socket, <<"Hello, Joe!\n":utf8>>)
let assert Ok(data) = mug.receive(socket, 500)
should.equal(data, <<"Hello, Joe!\n":utf8>>)
let assert Ok(_) = mug.shutdown(socket)
```

gdamjan commented on test/certs/gencerts.sh:
```diff
@@ -0,0 +1,34 @@
+  echo "Please run this script in the project root" >&2
+fi
+
+pushd test/certs
+if [[ -f server.crt || -f server.key || -f server.csr || -f ca.crt || -f ca.key ]]; then
+  popd
+  echo "A key already exists in test/keys. Proceeding will overwrite it!"
+  echo "Press ENTER to proceed, CTRL+C to cancel."
+  read
+  pushd test/certs
+fi
```
since this is mostly gonna be used on CI, I suggest you detect CI here and overwrite files anyway, otherwise just fail and make the user manually remove the certs.

Most (all?) CI systems export a `CI` environment variable?
what do you think?

ps.
I was inspired to look here by the somewhat unusual `yes | …` usage :D

gdamjan replied:
you also probably don't need the pushd/popd calls, since when you execute the script it wont modify the calling shells working directory

arnu515 replied:
Fixed.

lpil commented on test/mug_tls_test.gleam:
```diff
@@ -0,0 +1,175 @@
+
+  let assert Ok(mug.Packet(packet_socket, <<"Hello, Joe!\n":utf8>>)) =
+    process.select(selector, 1000)
+    |> io.debug()
```
Debug printing left in here!

arnu515 replied:
😬

lpil commented on test/mug_tls_test.gleam:
```diff
@@ -0,0 +1,175 @@
+
+  let assert Ok(_) = mug.shutdown(socket)
+
+  // if this sleep call does not exist, the below command *sometimes* errors out.
```
Why is this?

It looks like it should error, the pattern says it must.

arnu515 replied:
By "errors out" I meant the test fails. It's probably a little race condition that makes the `mug.send` function return an `inet:einval` error instead of a `mug.Closed` error as expected (this happens about 4/10 times). This probably has to do with how the SSL application is coded, I haven't looked into that yet.

That's why I added the sleep(1) call because even one millisecond is enough to give a predicatable output.

lpil commented on test/mug_tls_test.gleam:
```diff
@@ -0,0 +1,175 @@
+  let assert Ok(tcp_socket) =
+    mug.new("example.com", port: 443)
+    |> mug.connect()
+  let assert Ok(mug.SslSocket(socket)) =
+    mug.upgrade(tcp_socket, mug.DangerouslyDisableVerification, 5000)
+  let socket = mug.SslSocket(socket)
```
How come the socket is taken out and then put back into this record? Why are users of this library permitted to do that at all?

Also I think this test is supposed to be using the system CA, but it's disabling verification.

arnu515 replied:
Both issues have been fixed :+1:

lpil commented on test/mug_tls_test.gleam:
```diff
@@ -0,0 +1,175 @@
+
+pub const port = 64_794
+
+fn connect() {
```
Let's inline this function, looks like it is hardly used

arnu515 replied:
The `connect()` function? This function has 5 usages!

lpil commented on src/mug.gleam:
```diff
@@ -171,7 +396,162 @@ fn gen_tcp_connect(
+  case socket {
+    TcpSocket(socket) -> {
+      let opts =
+        erlang.rescue(fn() { get_tls_options(vm) })
```
Never use rescue like this, it silently discards actual problems. Return errors from the Erlang instead.

arnu515 replied:
Fixed!

lpil commented on src/mug.gleam:
```diff
@@ -1,20 +1,88 @@
+  /// The `ssl` application was not started/depended upon.
+  /// Either call `tls.start()` before calling any TLS methods, or (better yet) depend on the `ssl`
+  /// erlang application in [`gleam.toml`](https://gleam.run/writing-gleam/gleam-toml/).
+  SslNotStarted
```
Remove this please. It should never happen and the documentation gives incorrect advice.

arnu515 replied:
I had written this code before reading up on OTP behaviours 😅

lpil commented on src/mug.gleam:
```diff
@@ -1,20 +1,88 @@
+
+  // For connect only
+  /// An invalid option was passed
+  Options(opt: Dynamic)
```
What is this variant? It seems very unclear.

lpil replied:
This needs to be discussed 🙏

arnu515 replied:
Sorry, looks like I missed this. This is returned by `ssl:connect` if there is a mistake in the params list passed to it. This would (ideally) never happen, but I added it to the Error type so that if it occurs, it can get caught, instead of crashing the runtime with a match exception.

`ssl:connect`'s spec: https://www.erlang.org/doc/apps/ssl/ssl.html#connect/4 (see the definition of `Reason`).

lpil replied:
This needs to be refined to a proper error variant 🙏  We do not replicate Erlang APIs as-is in Gleam APIs.

arnu515 replied:
Well the docs do not specify a spec for it (just says `{options, any()}`) and I tried grepping through the source code to find all occurrences of this pattern, but the source is spread out over many files and I feel that I'll miss something or make a mistake.

This variant will only be hit if there is a problem with the parameters provided to the function, which will not happen since we have a safe wrapper. So, I feel that this error variant should be removed rather than type it out.

I don't feel that the effort to find every single possible type of `any()` is justified for something that will (ideally) never be hit.

lpil commented on src/mug.gleam:
```diff
@@ -1,20 +1,88 @@
+  /// Unable to get the OS supplied CA certificates. This error is only returned if the
+  /// `use_system_cacerts` option is set to `true` and the system's CA certificates
+  /// could not be retrieved.
+  SystemCacertificatesGetError
```
Is there no more information that can be given in this case?

lpil commented on src/mug.gleam:
```diff
@@ -109,13 +177,79 @@ pub type ConnectionOptions {
+  /// Path to a file containing PEM-encoded certificates and their key, with an optional
+  /// password associated with the file containing the key.
+  PemEncodedCertsKeys(
+    certfile: String,
+    keyfile: String,
```
```suggestion
    certificate_path: String,
    key_path: String,
```

lpil commented on src/mug.gleam:
```diff
@@ -109,13 +177,79 @@ pub type ConnectionOptions {
+pub type Key {
+  /// A der-encoded key.  
+  /// `alg` is one of 'RSAPrivateKey' | 'DSAPrivateKey' | 'ECPrivateKey' | 'PrivateKeyInfo'.
+  DerEncodedKey(alg: String, key: BitArray)
```
```suggestion
  DerEncodedKey(algorithm: String, key: BitArray)
```

If the string has to be one of a known finite variants then this should be a custom type rather than a string.

lpil commented on src/mug.gleam:
```diff
@@ -109,13 +177,79 @@ pub type ConnectionOptions {
+  Certificates(
+    use_system_cacerts: Bool,
+    cacerts: Option(CaCertificates),
+    certs_keys: List(CertsKeys),
```
```suggestion
    certificate_keys: List(CertsKeys),
```

arnu515 replied:
The erlang `ssl` app calls it `certs_keys`. We should follow the same.

lpil replied:
No we should not. Erlang/OTP typically has very hard to understand APIs, one of the main aims of Gleam is to improve on that.

karlsson replied:
Should not the type name be CertificateKeys also then? Or even CertificatesKeys?

arnu515 replied:
i've updated the field to be `certificates_keys`, and the type to be `CertificatesKeys`

lpil commented on src/mug.gleam:
```diff
@@ -109,13 +177,79 @@ pub type ConnectionOptions {
+  PemEncodedCaCertificates(cacertfile: String)
+}
+
+pub type CertsKeys {
```
This custom type is plural, but the data within is singular. Is this correct?

arnu515 replied:
That's because the certfile has a chain of certificates inside it.

lpil replied:
Sounds like it should be plural in that case. No abbreviations also please, we don't use them in Gleam and this branch currently mixes-and-matches full and abbreviations for the same thing.

arnu515 replied:
Changed this name to be ca_certificates_file

lpil commented on src/mug.gleam:
```diff
@@ -109,13 +177,79 @@ pub type ConnectionOptions {
+  /// certs override PEM-encoded ones. Therefore, if you wish to use a PEM encoded CA cert along
+  /// with the system's CA, you should decode the PEM into a DER using (`public_key:pem_decode`)[https://www.erlang.org/doc/apps/public_key/public_key.html#pem-api]
+  /// and use the DerEncodedCaCertificates variant instead.
+  Certificates(
```
It looks like one can use system certs wit custom certs, but the custom certs must all be der or pem. How come that specific restriction?

arnu515 replied:
If PEM certificates are passed, then the `ssl` application ignores DER certificates, including system cacerts, which are DER encoded. Thus if the user passes like a hundred DER certificates, passing just one PEM certificate will result in all those DER certs being ignored.

arnu515 replied:
This conversation is resolved, right?

lpil replied:
Yes, thank you!

lpil reviewed (commented):
Thank you! Great work here! Here's some notes:

The tests fail cryptically when run now due to certs being missing. The developer-experience needs to be good, so people don't get stuck there not knowing what to do. It would be good if it was automatic, or for the script being at the root (so it's more easy to run) and for instructions to be given to the programmer telling them what to do, instead of a cryptic crash.

The terms TLS and SSL are both used here, but it's not clear to me why either is picked in each case or what the difference is in this package. Could you expand on this a little please? 🙏  

Some of the functions are missing type annotations, could you ensure they all have all their annotations please.

For a few names abbreviations have been used and wordshavebeenwrittenlikethis. Please use full words always and put underscores between words, with the exception for existing established abbreviations or initialisms or acronyms such as TLS.

Remove all the use of catch all patterns (`_ -> ...`) please, they cause bugs when types change as the logic doesn't get updated.

The builder API can silently fail to configure TLS if you give it certificates but TLS has not been enabled, which seems like it could be error prone. Let's remove these builder functions and have the programmer explicitly construct the TLS configuration, to make it clear as possible.

And thank you for waiting while I dealt with some family matters.

AlexanderVanDamme commented:
hey, what's the status on this? in squirrel there's currently no way to connect to a database that requires ssl, so maybe this pr would make that possible?

karlsson commented:
> hey, what's the status on this? in squirrel there's currently no way to connect to a database that requires ssl, so maybe this pr would make that possible?

I [verified](https://github.com/giacomocavalieri/squirrel/issues/69#issuecomment-2790086481) that this PR makes it possible to use SSL in squirrel so it would be very nice if it could be approved eventually.

If I get it right the `mug.ConnectionOptions` gets a new parameter `tls_opts` which means changed API -> new major version?

lpil commented:
I agree! It would be great to get these issues resolved. If anyone would like to pick this up it would be much appreciated.

gideongrinberg commented:
> I agree! It would be great to get these issues resolved. If anyone would like to pick this up it would be much appreciated.

What changes need to be made before this is merged? It would be very helpful as many of the projects on the [SaaS wishlist](https://github.com/orgs/gleam-lang/projects/5/views/1) use SSL in one form or another.

arnu515 commented:
I had exams, projects, and other end-semester things in the month of April, spilling over to May, but that's all done and I'll be able to devote time to this PR! Apologies for no prior communication regarding this :sweat_smile: 

Can I bump deps like glisten because I'm getting quite a few deprecation warnings from the stdlib

arnu515 commented:
I'll have to remove a test for upgrading via system CA certificates, because most websites do not support connecting via TCP and then upgrading to TLS after.

It used to work for `example.com`, but it stopped working a few months ago. From looking at the logs, I assume that to perform this upgrade, a certificate for the IP address itself must be present, and not the domain name alone. Please correct me if I'm wrong here.

arnu515 commented:
All tests are passing now!

gideongrinberg commented:
@arnu515 thanks for doing this. I cloned your fork on my machine and get 7 failures. Any idea why that might be? 

```
......=NOTICE REPORT==== 16-May-2025::09:30:27.461927 ===
TLS client: In state wait_cert_cr at ssl_handshake.erl:2183 generated CLIENT ALERT: Fatal - Unknown CA

.=ERROR REPORT==== 16-May-2025::09:30:27.478337 ===
Failed to accept/start handler: ControlError
F.=ERROR REPORT==== 16-May-2025::09:30:27.486438 ===
Failed to accept/start handler: ControlError
F=ERROR REPORT==== 16-May-2025::09:30:27.490392 ===
Failed to accept/start handler: ControlError
F=ERROR REPORT==== 16-May-2025::09:30:27.494791 ===
Failed to accept/start handler: ControlError
F=ERROR REPORT==== 16-May-2025::09:30:27.498616 ===
Failed to accept/start handler: ControlError
F=ERROR REPORT==== 16-May-2025::09:30:27.502724 ===
Failed to accept/start handler: ControlError
F=ERROR REPORT==== 16-May-2025::09:30:27.506635 ===
Failed to accept/start handler: ControlError
F
Failures:

  1) mug_tls_test.connect_without_verification_test
     #{function => <<"connect">>,line => 12,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value => {error,closed},
       gleam_error => let_assert}
     location: mug_tls_test.connect:12
     stacktrace:
       mug_tls_test.connect
       mug_tls_test.connect_without_verification_test
     output: 

  2) mug_tls_test.upgrade_test
     #{function => <<"upgrade_test">>,line => 59,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value => {error,closed},
       gleam_error => let_assert}
     location: mug_tls_test.upgrade_test:59
     stacktrace:
       mug_tls_test.upgrade_test
     output: 

  3) mug_tls_test.upgrade_self_signed_test
     #{function => <<"upgrade_self_signed_test">>,line => 76,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value =>
           {error,
               {options,
                   {cacertfile,
                       "/Users/gideongrinberg/Documents/Coding/mug/test/certs/ca.crt",
                       {error,enoent}}}},
       gleam_error => let_assert}
     location: mug_tls_test.upgrade_self_signed_test:76
     stacktrace:
       mug_tls_test.upgrade_self_signed_test
     output: 

  4) mug_tls_test.hello_world_test
     #{function => <<"connect">>,line => 12,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value => {error,closed},
       gleam_error => let_assert}
     location: mug_tls_test.connect:12
     stacktrace:
       mug_tls_test.connect
       mug_tls_test.hello_world_test
     output: 

  5) mug_tls_test.active_mode_test
     #{function => <<"connect">>,line => 12,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value => {error,closed},
       gleam_error => let_assert}
     location: mug_tls_test.connect:12
     stacktrace:
       mug_tls_test.connect
       mug_tls_test.active_mode_test
     output: 

  6) mug_tls_test.exact_bytes_receive_test
     #{function => <<"connect">>,line => 12,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value => {error,closed},
       gleam_error => let_assert}
     location: mug_tls_test.connect:12
     stacktrace:
       mug_tls_test.connect
       mug_tls_test.exact_bytes_receive_test
     output: 

  7) mug_tls_test.exact_bytes_receive_not_enough_test
     #{function => <<"connect">>,line => 12,
       message => <<"Pattern match failed, no pattern matched the value.">>,
       module => <<"mug_tls_test">>,
       value => {error,closed},
       gleam_error => let_assert}
     location: mug_tls_test.connect:12
     stacktrace:
       mug_tls_test.connect
       mug_tls_test.exact_bytes_receive_not_enough_test
     output: 

Finished in 0.837 seconds
15 tests, 7 failures
```

lpil commented on src/mug.gleam:
```diff
@@ -3,18 +3,87 @@ import gleam/dynamic.{type Dynamic}
+}
+import mug/internal/system_cacerts
+
+pub type TcpSocket
```
Perhaps `ClearTextSocket`? As they're both TCP sockets.

I can't see anywhere this function is used in the public API. Perhaps it could be private?

lpil commented on src/mug.gleam:
```diff
@@ -3,18 +3,87 @@ import gleam/dynamic.{type Dynamic}
+}
 
-pub type Socket
+pub fn socket_is_ssl(socket: Socket) {
```
When might one use this function? If there's no use case let's remove it, otherwise add documentation please 🙏

lpil commented on src/mug.gleam:
```diff
@@ -3,18 +3,87 @@ import gleam/dynamic.{type Dynamic}
+import gleam/result
+import gleam/string
+import mug/internal/ssl_options.{
+  type SslOptionName, Cacertfile, Cacerts, CertsKeys, Verify,
```
Still have a mix of SSL and TLS terminology here. Which is more correct? I think it is TLS?

lpil commented on src/mug.gleam:
```diff
@@ -3,18 +3,87 @@ import gleam/dynamic.{type Dynamic}
+  Options(opt: Dynamic)
+
+  /// TLS connection failed
+  TlsAlert(alert: TlsAlert, description: String)
```
Let's have a test that asserts something about the description to ensure that it is a string

lpil reviewed (commented):
Great, thank you! There's a few comments that were not yet resolved, which I've highlighted with replies, and some items need documenting.

Thank you for your patience, I have got behind with my github notifications this release.

isaacharrisholt commented:
Just gonna mention that I've had a request for SSL support in Valkyrie (Pevensie/valkyrie#8), and having this in would greatly simplify that process. I'm happy to pick it up if you need any help @arnu515.

JesperWe commented:
Just want to add some cheers here (not skilled enough to contribute yet 😉) but TCP with TLS is used in many places. NATS is another protocol that would have use for this. Would be great if it could be part of mug.

karlsson commented:
Hi @arnu515 , I have squashed, rebased and adapted to latest version of mug in a [commit](https://github.com/karlsson/mug/tree/tls). I am not sure I have covered all outstanding questions but I think it will be a bit easier to proceed now.
Alternatively I could make a new PR from that.

arnu515 commented:
I have merged the main branch to this PR and updated the TLS code such that all tests are passing.

@karlsson I apologise I didn't see your commits or PR. Github didn't notify me for some reason.
I'd like us to merge our PRs into one. If you could send me a PR to my https://github.com/arnu515/mug repository with your commits, or give me permission to cherry-pick/merge your commits into my branch, that would be awesome.
Alternatively, we could continue development on your branch, if that's better.

I apologise for my inactivity in this PR, I kind of "forgot"/had other work to do.

karlsson commented:
Hi @arnu515 , since you are the original author I think you know best how to deal with this. Feel free to cherry pick whatever you want from my commits. I think the first one can be a bit tricky since it involved a squash of all your commits and then also adding code to adapt to latest rebase of mug / gleam_otp versions.
One thing that I [changed](https://github.com/karlsson/mug/commit/8a3f1c518c8e7489d4d0f8f53fe0dbeec6ec6ce4) was to generate CA and server certs with different CN, which makes using own CA to work. (no selfsigned_peer error). I don't know, maybe you would like to keep your certs and testcase as is, but I am not sure the OTP team will fix "pure" self-signed certs soon.

When diffing our branches arnu515/mug:main and karlsson/mug:tls one get a pretty good overview. Not too different :smile:.

arnu515 commented:
Thanks @karlsson ! I wasn't able to cherry-pick your commits due to too many conflicts, so I copy-pasted some of your code. I've added you as a Co-Author in those commits where I've taken your code verbatim. Thanks for the help! I've also integrated your changes with the cert generation and added CA cert checking to the tests, and all tests are passing. I do believe that we're not getting "pure" self-signed certs anytime soon from what I've seen on the mailing lists.

arnu515 commented:
Also I've left the gleam and dependency versions unchanged, unlike in @karlsson 's commits.

I'd like @lpil to decide on the minimum gleam and library versions himself. I've also not added a changelog nor bumped the (major) version of mug, so that needs to be done after the PR is merged!

arnu515 commented:
@isaacharrisholt @JesperWe I request you to test the current API by using my fork https://github.com/arnu515/mug and let me know if you have any suggestions!

lpil commented on src/mug.gleam:
```diff
@@ -3,12 +3,28 @@ import gleam/dynamic.{type Dynamic}
+  VerifyNone, VerifyPeer, combined_cert_into_cacerts, list_into_cacerts,
+}
+import mug/internal/system_cacerts
 
```
This module is used both the terms "SSL" and "TLS". Can we pick just one? Or is there a reason for the split?

lpil commented on src/mug.gleam:
```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
-/// to active mode and receive the next packet as an Erlang message.
+/// Uses the system's CA certificates to verify the server's certificate.
+///
+pub fn with_tls(options) {
```
Add missing annotations please.

None of the other functions have a `with_` prefix, let's rename this.

lpil commented on src/mug.gleam:
```diff
@@ -295,7 +411,79 @@ pub type ConnectionOptions {
     ///
     ip_version_preference: IpVersionPreference,
+    /// TLS options
+    tls_opts: TlsConnectionOptions,
```
No abbreviations please.

I think we could remove the wrapping here and make the configuration simpler?

```gleam
tls: Option(TlsVerification)
```
```gleam
pub type TlsVerification {
  DangerousUnverifiedTls
  VerifiedTls(use_system_cacerts: Bool, ...
}
```

lpil commented on src/mug.gleam:
```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
+
+/// Do not verify the server's certificate. This is dangerous and is not
+/// recommended. Use a custom certificate instead.
+pub fn dangerously_disable_verification(options) {
```
Remove this function please, I don't think we need a convenience for this.

lpil commented on src/mug.gleam:
```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
+///
+/// This is useful when you want to use your own CA certificates to verify the
+/// server's certificate. If verification is disabled, this function does nothing.
+pub fn no_system_cacerts(options) {
```
Remove this too please.

lpil commented on src/mug.gleam:
```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
 ///
+/// If verification is disabled, this function does nothing.
+///
+pub fn cacerts(
```
Remove this too please

lpil commented on src/mug.gleam:
```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
+///
+/// If verification is disabled, this function does nothing.
+///
+pub fn certificates_keys(
```
Remove this too please

lpil commented on src/mug.gleam:
```diff
@@ -354,67 +543,167 @@ pub type IpVersionPreference {
-type ModeValue {
-  Binary
+/// Returns the default gen_tcp options for mug
+fn get_tcp_options(use_inet6: Bool) -> List(GenTcpOption) {
```
Go back to the previous design of passing in the inet variant please, do not use a bool

lpil commented on src/mug.gleam:
```diff
@@ -424,13 +713,145 @@ type GenTcpOption {
+/// to perform the downgrade, which returns `ok` if the socket is closed, while
+/// this function returns an error. It is recommended against using this function
+/// to close the socket. Use `shutdown` instead.
+///
```
Remove details about the internal implementation from the API docs please.

lpil commented on src/mug.gleam:
```diff
@@ -538,9 +975,37 @@ pub fn select_tcp_messages(
+  fn(message) { mapper(unsafe_decode_tcp(message)) }
+}
+
+// TODO
```
TODO?

lpil commented on src/mug_ffi.erl:
```diff
@@ -1,31 +1,86 @@
 
--export([send/2, shutdown/1, coerce_tcp_message/1, active_once/0, passive/0]).
+-export([send/2, recv/3, shutdown/1, coerce_tcp_message/1, active_once/0, passive/0]).
+-export([ssl_upgrade/3, ssl_connect/5, ssl_downgrade/2, get_certs_keys/1, get_system_cacerts/0, coerce_ssl_message/1, coerce_unsafe/1]).
```
Wrap long lines and merge the two export statements into one please.

arnu515 replied:
It was to separate the TCP methods from the TLS methods but I'll merge them.

lpil commented on test/mug_test.gleam:
```diff
@@ -25,7 +26,15 @@ pub fn main() {
       glisten.continue(state)
     })
-    |> glisten.start(port)
+  // TODO: use new glisten methods
```
TODO?

lpil commented on src/mug_ffi.erl:
```diff
@@ -1,31 +1,86 @@
+    {socket_closed, {ssl_socket, Socket}};
+coerce_ssl_message({ssl_error, Socket, Error}) ->
+    {ssl_error, {ssl_socket, Socket}, Error}.
+coerce_unsafe(X) -> X.
```
Remove this please, no unsafe type casting in Gleam. Instead pass well typed data to Erlang FFI which then manipulates it as needed.

arnu515 replied:
I've updated these methods to thoroughly check their arguments.

lpil commented on src/mug_ffi.erl:
```diff
@@ -1,31 +1,86 @@
+recv({tcp_socket, Socket}, Length, Timeout) ->
+    gen_tcp:recv(Socket, Length, Timeout);
+recv({ssl_socket, Socket}, Length, Timeout) ->
+		ssl:recv(Socket, Length, Timeout).
```
Tabs being used for indentation here

lpil commented on src/mug_ffi.erl:
```diff
@@ -1,31 +1,86 @@
+send({tcp_socket, Socket}, Packet) ->
+    normalise(gen_tcp:send(Socket, Packet));
+send({ssl_socket, Socket}, Packet) ->
+		normalise(ssl:send(Socket, Packet)).
```
Tabs being used for indentation here

lpil commented on test/mug_tls_test.gleam:
```diff
@@ -0,0 +1,248 @@
+    |> mug.connect()
+
+  // This should crash with `badarg` if desc isn't a string
+  let _ = mug.describe_tls_alert(mug.UnknownCa) <> " - " <> desc
```
Use `assert mug.describe_tls_alert(mug.UnknownCa) == "some string"` please

lpil commented on test/mug_tls_test.gleam:
```diff
@@ -0,0 +1,248 @@
+// CA cert used for signing the server's cert
+pub const ca_crt = mug.PemEncodedCaCertificates("test/certs/ca.crt")
+
+// another CA cert that wasn't used for signing the server's cert
```
```suggestion
// Another CA cert that wasn't used for signing the server's cert
```

arnu515 commented:
I am having problems with getting github actions runners to run a test. It is not able to resolve example.com, gleam.run, or even github.com correctly.

The tests run successfully on my machine.

Apart from that, I have incorporated some your suggested changes and left comments on others.

karlsson commented on src/mug_ffi.erl:
```diff
@@ -1,31 +1,109 @@
+coerce_ssl_message({ssl_closed, Socket}) ->
+    {socket_closed, {ssl_socket, Socket}};
+coerce_ssl_message({ssl_error, Socket, Error}) ->
+    {ssl_error, {ssl_socket, Socket}, Error}.
```
Map to `tcp_error` ?
The `Message` type in mug.gleam only defines TcpError but the Error type covers TLS errors too.
```
coerce_ssl_message({ssl_error, Socket, Error}) ->
    {tcp_error, {ssl_socket, Socket}, Error}.
```

asumbek commented:
Is there going to be plan to add an option to set ALPN for a TLS socket?
